### PR TITLE
Vastly improve GPU support

### PR DESCRIPTION
**Pull request by mfherbst**
- Symmetries (via symmetrising on CPU)
- LDA and PBE
- Magnetic systems

### CODE REVIEW

**Comment by mfherbst** [comment]
@vchuravy Anything obvious comes to your mind, why the `Threads.@spawn` plus `@sync` leads to garbage when using CuArrays, but is completely fine with CPUs --- even for the case where `Threads.nthreads() == 1`?.

https://github.com/JuliaMolSim/DFTK.jl/blob/394d4ff4812a9bd6a1ab7bd51aabf042d0da3658/src/densities.jl#L39-L43

**Comment by vchuravy** [comment]
https://cuda.juliagpu.org/stable/usage/multitasking/#Task-based-programming

> so if we switch tasks and thus switch execution streams we need to synchronize() to ensure the data is actually available.

The docs are missing an example that clearly shows that at the end of a task you need to use a synchronizing operation.
Otherwise the CUDA ops are trailing and the task finishes to soon. So a `CUDA.synchronize()` at the end of the `@spawn` should work.

cc: @maleadt

File: src/workarounds/gpu_arrays.jl [hunk 1]
```diff
@@ -4,11 +4,10 @@ using GPUArraysCore
...
 
-function lowpass_for_symmetry!(ρ::AT, basis;
-    symmetries=basis.symmetries) where {AT <: AbstractGPUArray}
+function lowpass_for_symmetry!(ρ::AbstractGPUArray, basis; symmetries=basis.symmetries)
     all(isone, symmetries) && return ρ
     # lowpass_for_symmetry! currently uses scalar indexing, so we have to do this very ugly
     # thing for cases where ρ sits on a device (e.g. GPU)
     ρ_CPU = lowpass_for_symmetry!(to_cpu(ρ), basis; symmetries)
-    to_device(basis.architecture, ρ_CPU)
+    ρ .= to_device(basis.architecture, ρ_CPU)
```
Comment (antoine-levitt) [inline src/workarounds/gpu_arrays.jl:12]: I was going to say this is additional work in the case of CPU, but the compiler is smart enough to recognize that `x .= x` is a no-op (actually, it looks like it calls memmove directly, and memmove is smart enough to recognize the noop)

Comment (mfherbst) [reply src/workarounds/gpu_arrays.jl:12]: ... and additionally this is never called for CPU arrays by the way.

**Comment by mfherbst** [comment]
Thanks @vchuravy that makes sense.

**Comment by mfherbst** [comment]
Also @vchuravy and @maleadt: I suppose this synchronisation issue is present in all GPU backends, but I noticed there is no generic way to synchronise (like in `GPUArraysCore` or similar). Is this intentional or just a missing abstract function ?

**Comment by maleadt** [comment]
Different GPU back-ends have different notions of synchronization, so the abstraction would be confusing.

**Comment by mfherbst** [comment]
> Different GPU back-ends have different notions of synchronization, so the abstraction would be confusing.

Ok, that explains it, thanks.

**Comment by mfherbst** [comment]
Example run demonstrating the issue:
https://git.uni-paderborn.de/herbstm/DFTK.jl/-/jobs/129791#L655

**Pull request merged**